core/desugarModules.ml

```diff
@@ -17,328 +24,497 @@
-  object(self)
-    inherit SugarTraversals.fold_map as super
-
-    val term_shadow_table = term_ht
-    val type_shadow_table = type_ht
-    method get_term_shadow_table = term_shadow_table
-    method get_type_shadow_table = type_shadow_table
-    method bind_shadow_term name fqn =
-        {< term_shadow_table = shadow_binding name fqn term_shadow_table >}
-    method bind_shadow_type name fqn =
-        {< type_shadow_table = shadow_binding name fqn type_shadow_table >}
-
-    method! binder = function
-      | bndr ->
-         let n = Binder.to_name bndr in
-         let fqn = make_path_string path n in
-         (self#bind_shadow_term n fqn, Binder.set_name bndr fqn)
-
-    method! patternnode = function
-      | Pattern.Variant (n, p_opt) ->
-          let fqn = resolve n term_shadow_table in
-          let (o, p_opt') = self#option (fun o -> o#pattern) p_opt in
-          (o, Pattern.Variant (fqn, p_opt'))
-      | p -> super#patternnode p
-
-    method! row = function
-      | (xs, rv) ->
-          let (o, xs') =
-            self#list (fun o (name, fspec) ->
-              let (o, fspec') = o#fieldspec fspec in
-              (o, (name, fspec'))) xs in
-          let (_, rv') = o#row_var rv in
-          (self, (xs', rv'))
+  module Extend = struct
+    let module' : name -> t -> t -> t
+      = fun module_name module_scope scopes ->
+      { scopes with inner = S.Extend.module' module_name module_scope.inner scopes.inner }
+
+    let var : name -> string -> t -> t
+      = fun term_name prefixed_name scopes ->
+      { scopes with inner = S.Extend.var term_name prefixed_name scopes.inner }
+
+    let typename : name -> string -> t -> t
+      = fun typename prefixed_name scopes ->
+      { scopes with inner = S.Extend.typename typename prefixed_name scopes.inner }
+
+    let synthetic_module : name list -> scope -> t -> t
+      = fun path module_scope scopes ->
+      { scopes with inner = S.Extend.synthetic_module path module_scope scopes.inner }
+  end
+
+  let open_module : scope -> t -> t
+    = fun module_scope scopes ->
+    let inner = S.shadow scopes.inner module_scope in
+    { scopes with inner }
+
+  (* TODO. *)
+  let _import_module : name -> S.t -> t -> t
+    = fun _module_name _module_scope _scopes ->
+    assert false
+
+  let renew : t -> t
+    = fun scopes ->
+    let outer = S.shadow scopes.outer scopes.inner in
+    { outer; inner = S.empty }
+end
+
+let rec desugar_module : ?toplevel:bool -> Epithet.t -> Scope.t -> Sugartypes.binding -> binding list * Scope.t
+  = fun ?(toplevel=false) renamer scope binding ->
+  match binding.node with
+  | Module (name, bs) ->
+     let visitor = desugar ~toplevel (Epithet.remember ~escapes:(not toplevel) name renamer) (Scope.renew scope) in
+     let bs'    = visitor#bindings bs in
+     let scope' = visitor#get_scope in
+     let scope'' = Scope.Extend.module' name scope' scope in
+     (* Printf.printf "enter_module before:\n%s\n%!" (string_of_scopes scope);
```

Comment:
Can these comments go before merging?
(Same for the other comments below; maybe a better way of doing this would be to have debug log levels...)

Reply:
Oh yes, I will delete them. Thanks!